password/src/Data/Password/Validate.hs

```diff
@@ -104,37 +105,54 @@ isSpecial = \c ->
 -- | Character Category
 data CharacterCategory
   = Uppercase
+  -- ^ Uppercase letters
   | Lowercase
+  -- ^ Lowercase letters
   | Special
+  -- ^ Special characters
   | Digit
+  -- ^ ASCII digits
   deriving (Eq, Ord, Show)
 
+-- | Convert 'CharacterCategory' into associated predicate function
+categoryToPredicate :: CharacterCategory -> (Char -> Bool)
+categoryToPredicate = \case
+  Uppercase -> isAsciiUpper
+  Lowercase -> isAsciiLower
+  Special -> isSpecial
+  Digit -> isDigit
+
 -- | Possible reason of password being invalid
 data InvalidReason
   = PasswordTooShort !Int !Int
-  -- ^ Length of password is too short.
+  -- ^ Length of 'Password' is too short.
   --
   -- Expected at least 'Int' characters but the actual length is 'Int'
   | PasswordTooLong !Int !Int
-  -- ^ Length of password is too long.
+  -- ^ Length of 'Password' is too long.
   --
   -- Expected at maximum of 'Int' characters, but the actual length is 'Int'
   | NotEnoughReqChars CharacterCategory !Int !Int
-  -- ^ Password does not contain required number of characters.
+  -- ^ 'Password' does not contain required number of characters.
   --
   -- Expected at least 'Int' characters of 'CharacterCategory' but the password only
   -- contains 'Int'
-  | InvalidChar !Text
-  -- ^ Password contains characters that cannot be used
-  | InvalidPasswordPolicy !PasswordPolicy
-  -- ^ 'PasswordPolicy' is invalid
+  | InvalidCharacters !Text
+  -- ^ 'Password' contains characters that cannot be used
+  | InvalidLength !Int !Int
+  -- ^ Value of 'minimumLength' is bigger than 'maximumLength'
+  | MaxLengthBelowZero !Int
+  -- ^ Value of 'maximumLength' is less than zero
+  | InvalidCharSetPredicate !CharacterCategory !Int
```

Comment:
Added 'Int' value which indicates how many characters are required for a Password to be valid. This is for testing purposes.